bob-backend/src/pearl/group.rs

```diff
@@ -63,34 +64,42 @@ impl Group {
         }
     }
 
-    pub async fn run(&self, pp: impl Hooks) -> AnyResult<()> {
+    async fn run_under_reinit_lock(&self, pp: impl Hooks) -> AnyResult<()> {
         debug!("{}: read holders from disk", self);
         let config = self.settings.config();
-        let holders = config
+        let new_holders = config
             .try_multiple_times_async(
                 || self.read_vdisk_directory(),
                 "can't create pearl holders",
                 self.settings.config().fail_retry_timeout(),
             )
             .await
             .with_context(|| "backend pearl group read vdisk directory failed")?;
-        debug!("{}: count holders: {}", self, holders.len());
-        if !holders
+        debug!("{}: count holders: {}", self, new_holders.len());
+        if !new_holders
             .iter()
             .any(|holder| holder.gets_into_interval(get_current_timestamp()))
         {
             self.create_current_pearl();
         }
         debug!("{}: save holders to group", self);
-        self.add_range(holders).await;
+        {
+            let mut holders = self.holders.write().await;
+            holders.extend(new_holders).await;   
+        }
         debug!("{}: start holders", self);
         self.run_pearls(pp).await
```

Comment:
This is the only place, where `run_pearls` is called. I suggest to change `run_pearls` to not acquire the lock and pass `holders` as parameter. In this way the write lock taken here can be acquired up until the end of the function